### ISSUE TITLE
Unclear logic

### ISSUE THREAD

**Issue by planetA** [yaoxinjing/simpleDeviceDriver#9]
https://github.com/yaoxinjing/simpleDeviceDriver/blob/575ebb004b1771e8767032e12f068a7123ba28c3/device_file.c#L146-L155

Why would I want to pass `index==-1` to this function?

Maybe it helps if you explain what should happen in this function in plain English? And most importantly, motivate why it happens this way and not some other way.

**Comment by yaoxinjing** [comment]
from my point of view , wenn user want to write a key value paire to device file , the `device_file_write` function would first use 'key_exist' func to checke whether the key is already exist in the xarray, if the key exist , 'key_exist' func would return the index of the key value paire in the xarray, if the key not exist , the 'key_exist' func would return -1 as index.  then `device_file_write` will call  `key_value_store_save` function with the index(index of this key value paire in the xarray) and key value paire. in the `key_value_store_save`  func , it will check the index, if the index is -1, `key_value_store_save` fun would create a new item(key-value paire) in xarray, if the index is greater or equal than 0, it would replace the old item at the given index in xarray.

is that not a correct logic?

**Comment by planetA** [comment]
What is the difference between the "index" and the "key"?

**Comment by yaoxinjing** [comment]
the key is the offset of device file,and index is the location in xarray ,there store a pointer to a structure called key value paire. a key key value paire has key(offset of device file),pointer to value(data come from write system call),and the size of the data.

why i don't use offset of the device file as the index in xarray is that offset has type loff_t.  i didn't know whether "loff_t" is float or int type. therefore i decided to use a Conservative approach. in other world i created a struct called "key_value_pair" ,which inclued key(offset), pointer to value(data come from write system call),and the size of the data. and keep a global variable`next_index` to track the max index of xarray. 

everytime i need to store a new key value paire, i would store it to xarray[ next_index] .

**Comment by planetA** [comment]
For now, the way I read the code, you store all new elements with the index (key) taken from `next_index`. There is no way for the user to say that he want to save a key/value pair with a particular index. That is not how key/value stores work.

**Comment by planetA** [comment]
BTW, floating point numbers are generally forbidden in the kernel.

**Comment by yaoxinjing** [comment]
okay ,in the afternoon i will read some key value store example online and redisgn the logic of my project. 

maybe i could just use offset as key and index of the xarray. it would be more easier i suppose.
 
thanks :)

**Comment by planetA** [comment]
I think so too.
